File: py/plugins/middleware/src/genkit/plugins/middleware/_skills.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Skills middleware for Genkit.

Scans skill directories for SKILL.md files and injects a system prompt describing
available skills. Provides a use_skill tool for loading skill instructions into
the conversation context.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any, ClassVar

import yaml
from pydantic import Field, PrivateAttr

from genkit._core._model import GenerateHookParams, ModelResponse
from genkit.middleware import BaseMiddleware


class Skills(BaseMiddleware):
    """Skills middleware that exposes SKILL.md files as loadable instructions.

    Scans directories for subdirectories containing SKILL.md files. Each skill is
    exposed via a system prompt that lists available skills and their descriptions.
    The use_skill tool (when available) allows models to load full skill content.

    Skills are scanned once at initialization and cached for reuse across generate calls.
    """

    name: ClassVar[str] = 'middleware/skills'
    description: ClassVar[str | None] = 'Provides access to skill library for specialized instructions'

    skill_paths: list[str] = Field(default_factory=lambda: ['skills'])

    # Private cached state
    _skills_info: dict[str, dict[str, str]] | None = PrivateAttr(default=None)

    def _scan_skills(self) -> dict[str, dict[str, str]]:
        """Scan skill directories and return {skill_name: {path, description}}."""
        skills = {}

        for path_str in self.skill_paths:
            path = Path(path_str).resolve()
            if not path.is_dir():
                continue

            # List immediate subdirectories (non-hidden)
            for subdir in path.iterdir():
                if not subdir.is_dir() or subdir.name.startswith('.'):
                    continue

                skill_file = subdir / 'SKILL.md'
                if not skill_file.is_file():
                    continue

                # Parse frontmatter and extract description
                name, description = self._parse_skill_file(skill_file)
                if not name:
                    name = subdir.name

                skills[name] = {
                    'path': str(skill_file),
                    'description': description or 'No description provided.',
                }

        return skills

    def _parse_skill_file(self, path: Path) -> tuple[str, str]:
        """Parse SKILL.md frontmatter and return (name, description)."""
        try:
            content = path.read_text(encoding='utf-8')
        except Exception:
            return '', ''

        # Strip optional BOM
        content = content.lstrip('\ufeff')

        # Check for YAML frontmatter
        if not content.startswith('---\n'):
            return '', ''

        # Find end of frontmatter
        end_idx = content.find('\n---', 4)
        if end_idx == -1:
            return '', ''

        frontmatter = content[4:end_idx]
        try:
            data = yaml.safe_load(frontmatter)
            if not isinstance(data, dict):
                return '', ''
            return data.get('name', ''), data.get('description', '')
        except Exception:
            return '', ''

    def _get_skills_info(self) -> dict[str, dict[str, str]]:
        """Get or initialize cached skills info."""
        if self._skills_info is None:
            self._skills_info = self._scan_skills()
        return self._skills_info

    def _build_skills_prompt(self, info: dict[str, dict[str, str]]) -> str:
        """Build the skills system prompt from scanned info."""
        if not info:
            return ''

        lines = [
            '<skills>',
            'You have access to a library of skills that serve as specialized instructions/personas.',
            'Strongly prefer to use them when working on anything related to them.',
            'Only use them once to load the context.',
            'Here are the available skills:',
        ]

        for name in sorted(info.keys()):
            desc = info[name]['description']
            lines.append(f' - {name} - {desc}')

        lines.append('</skills>')
        return '\n'.join(lines)

    def _inject_skills_prompt(self, request: Any, prompt_text: str) -> Any:
        """Inject skills prompt into request as a system message part with marker metadata.

        Returns a shallow copy of the request with the prompt injected.
        """
        # Import here to avoid circular imports
        from genkit._core._typing import MessageData, Part, Role, TextPart

        # Make a shallow copy of the request
        request = request.model_copy(deep=False)

        # Find or create system message
        messages = list(request.messages)
        system_msg = None
        system_idx = None

        for i, msg in enumerate(messages):
            if msg.role == Role.SYSTEM:
                system_msg = msg
                system_idx = i
                break

        # Check if we already have the skills prompt
        marker_metadata = {'skills-instructions': True}

        if system_msg:
            # Check if identical prompt already exists
            for part in system_msg.content:
                if isinstance(part.root, TextPart):
                    if (
                        part.root.metadata
                        and part.root.metadata.get('skills-instructions')
                        and part.root.text == prompt_text
                    ):
                        # Already present, no change needed
                        return request

            # Add or update the skills prompt
            new_content = []
            found = False
            for part in system_msg.content:
                if isinstance(part.root, TextPart) and part.root.metadata and part.root.metadata.get('skills-instructions'):
                    # Replace existing skills prompt
                    new_content.append(Part(root=TextPart(text=prompt_text, metadata=marker_metadata)))
                    found = True
                else:
                    new_content.append(part)

            if not found:
                # Append new skills prompt
                new_content.append(Part(root=TextPart(text=prompt_text, metadata=marker_metadata)))

            # Update the system message
            system_msg = MessageData(role=Role.SYSTEM, content=new_content)
            messages[system_idx] = system_msg
        else:
            # Create a new system message
            part = Part(root=TextPart(text=prompt_text, metadata=marker_metadata))
            system_msg = MessageData(role=Role.SYSTEM, content=[part])
            messages.insert(0, system_msg)

        request.messages = messages
        return request

    async def wrap_generate(
        self,
        params: GenerateHookParams,
        next_fn,
    ) -> ModelResponse:
        """Inject skills prompt and use_skill tool (if possible) into the request."""
        info = self._get_skills_info()
        if not info:
            return await next_fn(params)

        # Build and inject skills prompt
        prompt_text = self._build_skills_prompt(info)
        if prompt_text:
            params.request = self._inject_skills_prompt(params.request, prompt_text)

        # TODO: Inject use_skill tool — requires wrap_generate tool injection API.
        # The use_skill tool should accept a skill_name string and return the full
        # SKILL.md content. Current middleware API does not provide a clean way to
        # inject tools into params.request.tools or params.options.tools from wrap_generate.
        # Once the API supports tool injection, implement:
        # - Tool definition: name='use_skill', input schema for skill_name
        # - Tool implementation: read and return skills_info[skill_name]['path'] content
        # - Inject into params.request or params.options

        return await next_fn(params)

File: js/plugins/middleware/src/skills.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as fs from 'fs';
import {
  X_GENKIT_ALLOW_CUSTOM,
  X_GENKIT_DATA_SOURCE,
  annotateSchema,
  generateMiddleware,
  z,
  type GenerateMiddleware,
} from 'genkit';
import { tool } from 'genkit/beta';
import * as path from 'path';

/**
 * Action that crawls the project directory for any directory containing a SKILL.md file.
 * This is used as a data source for the Dev UI.
 */
export const listSkillsAction = {
  name: 'folders/list',
  actionType: 'custom' as const,
  inputSchema: z.void(),
  outputSchema: z.array(
    z.object({
      label: z.string(),
      value: z.string(),
    })
  ),
  handler: async () => {
    const findSkillDirs = async (dir: string, results: string[] = []) => {
      try {
        const files = await fs.promises.readdir(dir);
        for (const file of files) {
          if (file === 'node_modules' || file === '.git' || file === 'lib')
            continue;
          const fullPath = path.join(dir, file);
          const stat = await fs.promises.stat(fullPath).catch(() => null);
          if (stat?.isDirectory()) {
            const hasSkillMd = await fs.promises
              .access(path.join(fullPath, 'SKILL.md'))
              .then(() => true)
              .catch(() => false);
            if (hasSkillMd) {
              results.push(path.relative(process.cwd(), fullPath));
            }
            await findSkillDirs(fullPath, results);
          }
        }
      } catch (e) {
        // Ignore errors for unreadable directories
      }
      return results;
    };
    const dirs = await findSkillDirs('.');
    return dirs.map((d) => ({ label: d, value: d }));
  },
};

export const SkillsOptionsSchema = z.object({
  /**
   * Paths to directories containing skills.
   * @default ['skills']
   */
  skillPaths: annotateSchema(
    z
      .array(z.string())
      .optional()
      .describe('Paths to directories containing skills.'),
    {
      [X_GENKIT_DATA_SOURCE]: '/custom/middleware:skills/folders/list',
      [X_GENKIT_ALLOW_CUSTOM]: true,
    }
  ),
});

export type SkillsOptions = z.infer<typeof SkillsOptionsSchema>;

/**
 * Creates a middleware that scans for skills in specified paths.
 * Injects a system prompt listing available skills and provides a `use_skill` tool.
 */
export const skills: GenerateMiddleware<typeof SkillsOptionsSchema> =
  generateMiddleware(
    {
      name: 'skills',
      description: 'Injects system instructions and tools for using skills.',
      configSchema: SkillsOptionsSchema,
    },
    ({ config }) => {
      const skillPaths = config?.skillPaths ?? ['skills'];
      const skillCache = new Map<
        string,
        { path: string; description: string }
      >();

      function parseFrontmatter(content: string) {
        const match = /^---\s*\r?\n([^]*?)\r?\n---/.exec(content);
        if (!match) return null;

        const yaml = match[1];
        const nameMatch = /^name:\s*(.+)/m.exec(yaml);
        const descriptionMatch = /^description:\s*(.+)/m.exec(yaml);

        return {
          name: nameMatch ? nameMatch[1].trim() : undefined,
          description: descriptionMatch
            ? descriptionMatch[1].trim()
            : undefined,
        };
      }

      let scanPromise: Promise<void> | null = null;

      function ensureSkillsScanned(): Promise<void> {
        if (!scanPromise) {
          scanPromise = (async () => {
            skillCache.clear();

            for (const p of skillPaths) {
              const dirPath = path.resolve(p);
              try {
                const files = await fs.promises.readdir(dirPath, {
                  withFileTypes: true,
                });
                for (const file of files) {
                  if (file.isDirectory() && !file.name.startsWith('.')) {
                    const skillDir = path.join(dirPath, file.name);
                    const skillMdPath = path.join(skillDir, 'SKILL.md');
                    try {
                      const content = await fs.promises.readFile(
                        skillMdPath,
                        'utf-8'
                      );
                      let description = 'No description provided.';
                      const fm = parseFrontmatter(content);
                      if (fm?.description) {
                        description = fm.description;
                      }
                      skillCache.set(file.name, {
                        path: skillMdPath,
                        description,
                      });
                    } catch (e) {
                      // ignore file read errors
                    }
                  }
                }
              } catch (e) {
                // ignore directory read errors
              }
            }
          })();
        }
        return scanPromise;
      }

      const useSkillTool = tool(
        {
          name: 'use_skill',
          description: 'Use a skill by its name.',
          inputSchema: z.object({
            skillName: z.string().describe('The name of the skill to use.'),
          }),
          outputSchema: z.string(),
        },
        async (input) => {
          await ensureSkillsScanned();
          const info = skillCache.get(input.skillName);
          if (!info) {
            throw new Error(`Skill '${input.skillName}' not found.`);
          }

          try {
            return await fs.promises.readFile(info.path, 'utf-8');
          } catch (e) {
            throw new Error(`Failed to read skill "${input.skillName}": ${e}`);
          }
        }
      );

      return {
        tools: [useSkillTool],
        generate: async (envelope, ctx, next) => {
          const { request } = envelope;
          await ensureSkillsScanned();
          if (skillCache.size === 0) return next(envelope, ctx);

          const skillsList = Array.from(skillCache.entries())
            .map(([name, info]) => {
              if (info.description !== 'No description provided.') {
                return ` - ${name} - ${info.description}`;
              }
              return ` - ${name}`;
            })
            .join('\n');

          const systemPromptText =
            `<skills>\n` +
            `You have access to a library of skills that serve as specialized instructions/personas.\n` +
            `Strongly prefer to use them when working on anything related to them.\n` +
            `Only use them once to load the context.\n` +
            `Here are the available skills:\n` +
            `${skillsList}\n` +
            `</skills>`;

          const messages = [...request.messages];
          let injectedPart: any | undefined;
          let injectedMsgIndex = -1;
          let injectedPartIndex = -1;

          for (let i = 0; i < messages.length; i++) {
            const msg = messages[i];
            for (let j = 0; j < msg.content.length; j++) {
              const p = msg.content[j];
              if (p.text && p.metadata?.['skills-instructions'] === true) {
                injectedPart = p;
                injectedMsgIndex = i;
                injectedPartIndex = j;
                break;
              }
            }
            if (injectedPart) break;
          }

          if (injectedPart) {
            if (injectedPart.text !== systemPromptText) {
              const newContent = [...messages[injectedMsgIndex].content];
              newContent[injectedPartIndex] = {
                text: systemPromptText,
                metadata: { 'skills-instructions': true },
              };
              messages[injectedMsgIndex] = {
                ...messages[injectedMsgIndex],
                content: newContent as any,
              };
            }
          } else {
            const systemMsgIndex = messages.findIndex(
              (m) => m.role === 'system'
            );
            if (systemMsgIndex !== -1) {
              messages[systemMsgIndex] = {
                ...messages[systemMsgIndex],
                content: [
                  ...messages[systemMsgIndex].content,
                  {
                    text: systemPromptText,
                    metadata: { 'skills-instructions': true },
                  },
                ],
              };
            } else {
              messages.unshift({
                role: 'system',
                content: [
                  {
                    text: systemPromptText,
                    metadata: { 'skills-instructions': true },
                  },
                ],
              });
            }
          }

          return next({ ...envelope, request: { ...request, messages } }, ctx);
        },
      };
    }
  );

// Add the discovery action to the plugin's init hook
const originalPlugin = skills.plugin;
skills.plugin = (options) => {
  const p = originalPlugin(options);
  const originalInit = p.init;
  p.init = async () => {
    const { action } = require('genkit');
    const supplementalActions = [
      action(listSkillsAction as any, listSkillsAction.handler),
    ];
    if (originalInit) {
      const originalActions = await originalInit();
      return [...(originalActions || []), ...supplementalActions];
    }
    return supplementalActions;
  };
  return p;
};
